File: src/tokamak/pipeline.py
"""End-to-end orchestration. CLI is a thin wrapper over `run()`.

For each row (trace) the pipeline runs:

  1. Find reasoning spans + surrounding problem/answer context.
  2. Process each span: compress (terse rewrite), invert (skeleton ->
     fuller trace), or both (compress then invert). Modes can also be
     `rules` (regex-only) or `noop` (passthrough).
  3. QAQC: an independent validation agent grades the processed span on
     0..1. Optional `mirror_qaqc > 1` runs N judges in parallel and keeps
     the worst score (one bad reading dominates).
  4. Apply processed text back to the trace.
  5. Anonymize, dedup, score, classify, export.

The per-trace `signal` (average span signal) is written into the row's
`metadata.signal` and into a dedicated `signal` column in the
compression report.
"""

from __future__ import annotations

import hashlib
import json
import logging
from collections import Counter
from dataclasses import dataclass, field
from pathlib import Path
from typing import Any, Dict, List, Optional, Tuple

from . import anonymize, card, classify, dedup, export, extract, prompts
from .caveman import estimate_tokens
from .processor import Processor, ProcessResult, SpanInput
from .validator import ValidationResult, Validator

logger = logging.getLogger("tokamak")


# ---------------------------------------------------------------------------
# Loading
# ---------------------------------------------------------------------------

def _load_jsonl(path: Path) -> List[Dict[str, Any]]:
    out: List[Dict[str, Any]] = []
    with path.open("r", encoding="utf-8") as f:
        for i, line in enumerate(f, 1):
            line = line.strip()
            if not line:
                continue
            try:
                out.append(json.loads(line))
            except json.JSONDecodeError as exc:
                logger.warning("%s:%d bad JSON: %s", path.name, i, exc)
    return out


def load_traces(input_dir: Optional[Path], input_file: Optional[Path]) -> List[Dict[str, Any]]:
    traces: List[Dict[str, Any]] = []
    if input_file:
        traces.extend(_load_jsonl(input_file))
    if input_dir:
        for p in sorted(input_dir.rglob("*.jsonl")):
            traces.extend(_load_jsonl(p))
    return traces


# ---------------------------------------------------------------------------
# Stats
# ---------------------------------------------------------------------------

@dataclass
class TraceStats:
    spans: int = 0
    original_tokens: int = 0
    processed_tokens: int = 0
    signal_sum: float = 0.0
    signal_count: int = 0

    @property
    def signal(self) -> float:
        if self.signal_count == 0:
            return 0.0
        return self.signal_sum / self.signal_count


@dataclass
class RunStats:
    traces_in: int = 0
    traces_out: int = 0
    spans: int = 0
    original_tokens: int = 0
    compressed_tokens: int = 0   # name kept for backward-compat with card.render
    redactions: int = 0
    avg_signal: float = 0.0
    per_trace: List[Dict[str, Any]] = field(default_factory=list)

    def as_dict(self) -> Dict[str, Any]:
        return {
            "traces_in": self.traces_in,
            "traces_out": self.traces_out,
            "spans": self.spans,
            "original_tokens": self.original_tokens,
            "compressed_tokens": self.compressed_tokens,
            "redactions": self.redactions,
            "avg_signal": self.avg_signal,
        }


# ---------------------------------------------------------------------------
# Stage runners
# ---------------------------------------------------------------------------

def _collect_spans(
    traces: List[Dict[str, Any]]
) -> Tuple[List[extract.Span], List[int]]:
    """Flatten spans across all traces. Returns (spans, trace_index_per_span)."""
    flat: List[extract.Span] = []
    owner: List[int] = []
    for ti, trace in enumerate(traces):
        for sp in extract.find_reasoning_spans(trace):
            flat.append(sp)
            owner.append(ti)
    return flat, owner


def _to_inputs(spans: List[extract.Span]) -> List[SpanInput]:
    return [
        SpanInput(problem=sp.problem, answer=sp.answer, reasoning=sp.text)
        for sp in spans
    ]


def _run_stage(
    spans: List[extract.Span],
    processor: Processor,
    seqs: int,
) -> List[ProcessResult]:
    inputs = _to_inputs(spans)
    return processor.process_batch(inputs, max_workers=max(1, seqs))


# ---------------------------------------------------------------------------
# Top-level
# ---------------------------------------------------------------------------

def _trace_id(trace: Dict[str, Any], index: int) -> str:
    raw = json.dumps(trace, sort_keys=True, ensure_ascii=False).encode("utf-8")
    return f"t{index:06d}_{hashlib.sha1(raw).hexdigest()[:10]}"


def _apply_signal(trace: Dict[str, Any], signal: float) -> None:
    """Stamp `signal` into a stable place on the row for downstream tooling."""
    md = trace.setdefault("metadata", {})
    if not isinstance(md, dict):
        # Trace already used `metadata` for something else — keep both.
        md = {}
        trace["metadata"] = md
    md["signal"] = round(signal, 4)
    trace["signal"] = round(signal, 4)


def run(
    *,
    input_dir: Optional[Path],
    input_file: Optional[Path],
    output_dir: Path,
    compress_mode: str = "rules",
    caveman_level: str = "lite",       # back-compat: maps to processor `level`
    anonymize_level: str = "strict",
    max_similarity: float = 0.92,
    model: Optional[str] = None,
    llm_concurrency: int = 1,          # back-compat alias for `seqs`
    seqs: Optional[int] = None,
    process_mode: Optional[str] = None,  # "compress" | "invert" | "both" | "rules" | "noop"
    level: Optional[str] = None,
    qaqc: bool = False,
    qaqc_mirrors: int = 1,
) -> RunStats:
    """Run the full pipeline. See module docstring for stage order.

    Backward-compat: callers using `compress_mode` / `caveman_level` /
    `llm_concurrency` still work — they map onto the new processor mode,
    level, and concurrency parameters.
    """
    traces = load_traces(input_dir, input_file)
    if not traces:
        raise SystemExit("no traces loaded")

    output_dir.mkdir(parents=True, exist_ok=True)
    (output_dir / "prompts").mkdir(exist_ok=True)
    (output_dir / "prompts" / "compress_lite.md").write_text(prompts.compress_prompt("lite"))
    (output_dir / "prompts" / "compress_full.md").write_text(prompts.compress_prompt("full"))
    (output_dir / "prompts" / "invert.md").write_text(prompts.invert_prompt())
    (output_dir / "prompts" / "validate.md").write_text(prompts.validate_prompt())
    # Back-compat: keep the old filenames around so existing consumers don't break.
    (output_dir / "prompts" / "caveman_lite.md").write_text(prompts.compress_prompt("lite"))
    (output_dir / "prompts" / "caveman_full.md").write_text(prompts.compress_prompt("full"))

    # Resolve effective config.
    eff_mode = (process_mode or compress_mode or "rules").lower()
    if eff_mode == "llm":
        eff_mode = "compress"          # back-compat alias
    eff_level = (level or caveman_level or "lite").lower()
    eff_seqs = seqs if seqs is not None else max(1, llm_concurrency)

    if eff_mode not in ("rules", "compress", "invert", "both", "noop"):
        raise ValueError(f"unknown process_mode: {eff_mode!r}")

    # Build processors and (optional) validator.
    stages: List[Processor]
    if eff_mode == "both":
        stages = [
            Processor("compress", level=eff_level, model=model),
            Processor("invert", model=model),
        ]
    else:
        stages = [Processor(eff_mode, level=eff_level, model=model)]

    validator = Validator(model=model, mirrors=qaqc_mirrors) if qaqc else None

    # Flatten spans.
    flat_spans, owner = _collect_spans(traces)

    stats = RunStats(traces_in=len(traces))
    per_trace_stats: List[TraceStats] = [TraceStats() for _ in traces]

    # Capture originals before any stage rewrites them — needed for QAQC.
    originals: List[SpanInput] = _to_inputs(flat_spans)

    # Sequential stages: compress → (optional) invert. Each stage rebatches.
    current_results: Optional[List[ProcessResult]] = None
    for processor in stages:
        # For the *next* stage, feed the previous stage's output as the
        # reasoning input — same problem/answer rails, new reasoning text.
        if current_results is not None:
            for sp, pr in zip(flat_spans, current_results):
                sp.text = pr.processed
        current_results = _run_stage(flat_spans, processor, eff_seqs)

    final_results = current_results or []

    # QAQC: validate each final-processed span vs its captured original.
    signals: List[Optional[ValidationResult]] = [None] * len(flat_spans)
    if validator is not None and final_results:
        from concurrent.futures import ThreadPoolExecutor

        def _grade(i: int) -> Tuple[int, ValidationResult]:
            orig_input = originals[i]
            mode_for_judge = "invert" if stages[-1].mode == "invert" else (
                "compress" if stages[-1].mode in ("compress", "rules") else "compress"
            )
            return i, validator.validate(
                orig_input, final_results[i].processed, mode_for_judge
            )

        if eff_seqs > 1:
            with ThreadPoolExecutor(max_workers=eff_seqs) as ex:
                for i, vr in ex.map(_grade, range(len(flat_spans))):
                    signals[i] = vr
        else:
            for i in range(len(flat_spans)):
                _, vr = _grade(i)
                signals[i] = vr

    # Apply final processed text back to the trace and roll up per-trace stats.
    for i, (sp, pr) in enumerate(zip(flat_spans, final_results)):
        sp.apply(pr.processed)
        ti = owner[i]
        ts = per_trace_stats[ti]
        ts.spans += 1
        ts.original_tokens += estimate_tokens(originals[i].reasoning)
        ts.processed_tokens += pr.processed_tokens
        if signals[i] is not None:
            ts.signal_sum += signals[i].signal  # type: ignore[union-attr]
            ts.signal_count += 1

    # Anonymize and stamp signal onto each row.
    processed_traces: List[Dict[str, Any]] = []
    signal_values: List[float] = []
    for ti, trace in enumerate(traces):
        ts = per_trace_stats[ti]
        stats.spans += ts.spans
        stats.original_tokens += ts.original_tokens
        stats.compressed_tokens += ts.processed_tokens

        trace, n_redactions = anonymize.redact_trace(trace, level=anonymize_level)
        stats.redactions += n_redactions

        # Default signal = 1.0 if QAQC disabled (we don't have a reading;
        # mark as unknown via -1.0 so consumers can distinguish).
        if validator is None:
            row_signal = -1.0
        else:
            row_signal = ts.signal if ts.signal_count else -1.0
        _apply_signal(trace, row_signal)
        if row_signal >= 0:
            signal_values.append(row_signal)

        stats.per_trace.append({
            "index": ti,
            "spans": ts.spans,
            "tokens_before": ts.original_tokens,
            "tokens_after": ts.processed_tokens,
            "redactions": n_redactions,
            "signal": row_signal,
        })
        processed_traces.append(trace)

    stats.avg_signal = (
        sum(signal_values) / len(signal_values) if signal_values else 0.0
    )

    # Dedup.
    deduped, dropped = dedup.dedup(processed_traces, max_similarity=max_similarity)
    logger.info("dedup dropped %d / %d", dropped, len(processed_traces))

    # Score + classify + format.
    from .quality import score as quality_score
    axolotl_lines: List[str] = []
    sharegpt_lines: List[str] = []
    unsloth_lines: List[str] = []
    score_values: List[float] = []
    error_counts: Counter = Counter()

    for i, trace in enumerate(deduped):
        composite, _breakdown = quality_score(trace)
        err = classify.classify(trace)
        tid = _trace_id(trace, i)

        score_values.append(composite)
        error_counts[err] += 1

        axolotl_lines.append(json.dumps(
            export.axolotl(trace, composite, tid, err), ensure_ascii=False
        ))
        sharegpt_lines.append(json.dumps(export.sharegpt(trace), ensure_ascii=False))
        unsloth_lines.append(json.dumps(
            export.unsloth(trace, composite, tid, err), ensure_ascii=False
        ))

    (output_dir / "data.jsonl").write_text(
        "\n".join(axolotl_lines) + "\n", encoding="utf-8"
    )
    (output_dir / "sharegpt.jsonl").write_text(
        "\n".join(sharegpt_lines) + "\n", encoding="utf-8"
    )
    (output_dir / "unsloth.jsonl").write_text(
        "\n".join(unsloth_lines) + "\n", encoding="utf-8"
    )

    stats.traces_out = len(deduped)

    (output_dir / "dataset_card.md").write_text(
        card.render(stats.as_dict(), score_values, error_counts), encoding="utf-8"
    )

    # Per-trace processing report. `signal` column shown alongside compression.
    report = [
        "# Processing report",
        "",
        "| trace | spans | tokens_before | tokens_after | saved % | signal |",
        "|-------|-------|---------------|--------------|---------|--------|",
    ]
    for row in stats.per_trace:
        before = row["tokens_before"] or 0
        after = row["tokens_after"] or 0
        saved = 0.0 if before == 0 else 100.0 * (1 - after / before)
        sig = row["signal"]
        sig_cell = "—" if sig < 0 else f"{sig:.3f}"
        report.append(
            f"| {row['index']} | {row['spans']} | {before} | {after} | "
            f"{saved:.1f} | {sig_cell} |"
        )
    (output_dir / "compression_report.md").write_text(
        "\n".join(report) + "\n", encoding="utf-8"
    )

    return stats


def push_to_hub(out_dir: Path, repo_id: str, private: bool = True) -> None:
    try:
        from huggingface_hub import HfApi  # type: ignore
    except ImportError as exc:
        raise SystemExit("install `huggingface_hub` to use --push-to-hub") from exc
    api = HfApi()
    api.create_repo(repo_id, repo_type="dataset", private=private, exist_ok=True)
    api.upload_folder(folder_path=str(out_dir), repo_id=repo_id, repo_type="dataset")
    logger.info("pushed to https://huggingface.co/datasets/%s", repo_id)

File: rust/tokamak/src/pipeline.rs
//! Top-level orchestrator. Loads JSONL, finds spans, runs processor stages,
//! optionally validates, anonymizes, scores, dedups, exports.

use std::collections::BTreeMap;
use std::fs;
use std::io::{BufRead, BufReader, Write};
use std::path::{Path, PathBuf};

use anyhow::{Context, Result};
use serde_json::{json, Value};

use crate::anonymize::{self, Level as AnonLevel};
use crate::card::{render_card, render_report, CardStats, PerTraceRow};
use crate::classify;
use crate::dedup;
use crate::export;
use crate::extract::{self, Span};
use crate::llm::{LlmClient, LlmConfig};
use crate::processor::{Mode, Processor, ProcessResult, SpanInput};
use crate::prompts::{self, Level};
use crate::quality;
use crate::terse::estimate_tokens;
use crate::validator::Validator;

#[derive(Debug, Clone)]
pub struct RunOptions {
    pub input_dir:       Option<PathBuf>,
    pub input_file:      Option<PathBuf>,
    pub output_dir:      PathBuf,
    pub mode:            String,         // "rules" | "compress" | "invert" | "both" | "noop"
    pub level:           Level,
    pub anonymize_level: AnonLevel,
    pub max_similarity:  f64,
    pub seqs:            usize,
    pub qaqc:            bool,
    pub qaqc_mirrors:    usize,
    pub endpoint:        Option<String>,
    pub model:           Option<String>,
}

#[derive(Debug, Default, serde::Serialize)]
pub struct RunStats {
    pub traces_in:        usize,
    pub traces_out:       usize,
    pub spans:            usize,
    pub original_tokens:  usize,
    pub compressed_tokens:usize,
    pub redactions:       usize,
    pub avg_signal:       f64,
}

pub fn load_traces(opts: &RunOptions) -> Result<Vec<Value>> {
    let mut traces = Vec::new();
    if let Some(p) = &opts.input_file { traces.extend(load_jsonl(p)?); }
    if let Some(d) = &opts.input_dir  {
        for entry in walk_jsonl(d)? { traces.extend(load_jsonl(&entry)?); }
    }
    Ok(traces)
}

fn load_jsonl(path: &Path) -> Result<Vec<Value>> {
    let f = fs::File::open(path).with_context(|| format!("open {}", path.display()))?;
    let r = BufReader::new(f);
    let mut out = Vec::new();
    for (i, line) in r.lines().enumerate() {
        let line = line?;
        if line.trim().is_empty() { continue; }
        match serde_json::from_str::<Value>(&line) {
            Ok(v) => out.push(v),
            Err(e) => tracing::warn!("{}:{} bad JSON: {}", path.display(), i+1, e),
        }
    }
    Ok(out)
}

fn walk_jsonl(dir: &Path) -> Result<Vec<PathBuf>> {
    let mut out = Vec::new();
    for entry in fs::read_dir(dir)? {
        let entry = entry?;
        let path = entry.path();
        if path.is_dir() { out.extend(walk_jsonl(&path)?); }
        else if path.extension().and_then(|x| x.to_str()) == Some("jsonl") { out.push(path); }
    }
    out.sort();
    Ok(out)
}

fn stages_for(mode: &str) -> Result<Vec<Mode>> {
    match mode {
        "rules"    => Ok(vec![Mode::Rules]),
        "compress" => Ok(vec![Mode::Compress]),
        "invert"   => Ok(vec![Mode::Invert]),
        "both"     => Ok(vec![Mode::Compress, Mode::Invert]),
        "noop"     => Ok(vec![Mode::Noop]),
        other      => anyhow::bail!("unknown mode: {other:?}"),
    }
}

#[tokio::main]
pub async fn run(opts: RunOptions) -> Result<RunStats> {
    let traces = load_traces(&opts)?;
    if traces.is_empty() { anyhow::bail!("no traces loaded"); }

    fs::create_dir_all(&opts.output_dir)?;
    fs::create_dir_all(opts.output_dir.join("prompts"))?;
    write_prompts(&opts.output_dir)?;

    let stages = stages_for(&opts.mode)?;
    let needs_llm = stages.iter().any(|m| matches!(m, Mode::Compress | Mode::Invert)) || opts.qaqc;

    // Build client once, shared across stages.
    let llm = if needs_llm {
        Some(LlmClient::new(LlmConfig::from_env_or(opts.endpoint.clone(), opts.model.clone()))
            .context("build llm client")?)
    } else { None };

    // Mutable trace store + flat span list.
    let mut traces = traces;
    let (flat_spans, owner) = collect_spans(&traces);

    // Capture originals before any rewrite (needed for QAQC).
    let originals: Vec<SpanInput> = flat_spans.iter()
        .map(|s| SpanInput { problem: s.problem.clone(), answer: s.answer.clone(), reasoning: s.text.clone() })
        .collect();

    // Run each stage in order. Each stage produces new reasoning text for
    // every span; the next stage feeds it as input.
    let mut current_reasoning: Vec<String> = flat_spans.iter().map(|s| s.text.clone()).collect();
    let mut final_results: Vec<ProcessResult> = Vec::new();
    let mut final_mode = Mode::Noop;

    for stage_mode in stages.iter().copied() {
        let proc = Processor::new(stage_mode, opts.level);
        let inputs: Vec<SpanInput> = flat_spans.iter().enumerate().map(|(i, s)| SpanInput {
            problem: s.problem.clone(), answer: s.answer.clone(),
            reasoning: current_reasoning[i].clone(),
        }).collect();

        let results = match stage_mode {
            Mode::Rules | Mode::Noop => inputs.iter().map(|s| proc.process_local(s)).collect(),
            Mode::Compress | Mode::Invert => {
                proc.process_batch_llm(llm.clone().expect("llm client"), &inputs, opts.seqs.max(1)).await
            }
        };

        for (i, r) in results.iter().enumerate() {
            current_reasoning[i] = r.processed.clone();
        }
        final_results = results;
        final_mode = stage_mode;
    }

    // QAQC pass.
    let mut signals: Vec<Option<f64>> = vec![None; flat_spans.len()];
    let mut avg_signal_sum = 0.0;
    let mut avg_signal_count = 0usize;
    if opts.qaqc && !flat_spans.is_empty() {
        let v = Validator::new(opts.qaqc_mirrors);
        let mode_name: &'static str = match final_mode {
            Mode::Invert => "invert",
            _            => "compress",
        };
        let items: Vec<(SpanInput, String, &'static str)> = originals.iter().enumerate()
            .map(|(i, orig)| (orig.clone(), final_results[i].processed.clone(), mode_name))
            .collect();
        let grades = v.validate_batch(llm.clone().expect("llm client"), items, opts.seqs.max(1)).await;
        for (i, g) in grades.into_iter().enumerate() {
            signals[i] = Some(g.signal);
            avg_signal_sum += g.signal;
            avg_signal_count += 1;
        }
    }

    // Apply final processed text back into the trace JSON.
    for (i, span) in flat_spans.iter().enumerate() {
        extract::apply_replacement(&mut traces[owner[i]], span, &final_results[i].processed);
    }

    // Per-trace rollup.
    let mut per_trace_rows: Vec<PerTraceRow> = (0..traces.len()).map(|i| PerTraceRow {
        index: i, spans: 0, tokens_before: 0, tokens_after: 0, redactions: 0, signal: -1.0,
    }).collect();
    let mut per_trace_signal_sum: Vec<f64> = vec![0.0; traces.len()];
    let mut per_trace_signal_n: Vec<usize> = vec![0; traces.len()];
    for (i, _span) in flat_spans.iter().enumerate() {
        let ti = owner[i];
        per_trace_rows[ti].spans += 1;
        per_trace_rows[ti].tokens_before += estimate_tokens(&originals[i].reasoning);
        per_trace_rows[ti].tokens_after  += final_results[i].processed_tokens;
        if let Some(s) = signals[i] {
            per_trace_signal_sum[ti] += s;
            per_trace_signal_n[ti] += 1;
        }
    }

    // Anonymize each trace, stamp signal, accumulate redactions.
    let mut redactions = 0usize;
    for (ti, trace) in traces.iter_mut().enumerate() {
        let n = anonymize::redact_trace(trace, opts.anonymize_level);
        redactions += n;
        per_trace_rows[ti].redactions = n;
        let signal = if per_trace_signal_n[ti] > 0 {
            per_trace_signal_sum[ti] / per_trace_signal_n[ti] as f64
        } else { -1.0 };
        per_trace_rows[ti].signal = signal;
        stamp_signal(trace, signal);
    }

    // Dedup.
    let (kept, dropped) = dedup::dedup(traces, opts.max_similarity);
    tracing::info!("dedup dropped {} / {}", dropped, kept.len() + dropped);

    // Score + classify + export.
    let mut score_values: Vec<f64> = Vec::with_capacity(kept.len());
    let mut error_counts: BTreeMap<&str, usize> = BTreeMap::new();
    let axolotl_path  = opts.output_dir.join("data.jsonl");
    let sharegpt_path = opts.output_dir.join("sharegpt.jsonl");
    let unsloth_path  = opts.output_dir.join("unsloth.jsonl");
    let mut axo_f = fs::File::create(&axolotl_path)?;
    let mut shg_f = fs::File::create(&sharegpt_path)?;
    let mut uns_f = fs::File::create(&unsloth_path)?;

    for (i, trace) in kept.iter().enumerate() {
        let sc = quality::score(trace);
        let err = classify::classify(trace);
        let tid = trace_id(trace, i);
        score_values.push(sc.composite);
        *error_counts.entry(err).or_insert(0) += 1;

        writeln!(axo_f, "{}", export::axolotl(trace, sc.composite, &tid, err))?;
        writeln!(shg_f, "{}", export::sharegpt(trace))?;
        writeln!(uns_f, "{}", export::unsloth(trace, sc.composite, &tid, err))?;
    }

    let stats = RunStats {
        traces_in:        per_trace_rows.len(),
        traces_out:       kept.len(),
        spans:            flat_spans.len(),
        original_tokens:  per_trace_rows.iter().map(|r| r.tokens_before).sum(),
        compressed_tokens:per_trace_rows.iter().map(|r| r.tokens_after).sum(),
        redactions,
        avg_signal: if avg_signal_count > 0 { avg_signal_sum / avg_signal_count as f64 } else { 0.0 },
    };

    let card_stats = CardStats {
        traces_in: stats.traces_in, traces_out: stats.traces_out, spans: stats.spans,
        original_tokens: stats.original_tokens, processed_tokens: stats.compressed_tokens,
        redactions: stats.redactions, avg_signal: stats.avg_signal,
    };
    fs::write(opts.output_dir.join("dataset_card.md"),
              render_card(&card_stats, &score_values, &error_counts, ""))?;
    fs::write(opts.output_dir.join("compression_report.md"),
              render_report(&per_trace_rows))?;

    Ok(stats)
}

fn stamp_signal(trace: &mut Value, signal: f64) {
    if let Value::Object(o) = trace {
        let md = o.entry("metadata".to_string()).or_insert_with(|| json!({}));
        if let Value::Object(mdo) = md {
            mdo.insert("signal".into(), json!((signal * 10000.0).round() / 10000.0));
        }
        o.insert("signal".into(), json!((signal * 10000.0).round() / 10000.0));
    }
}

fn collect_spans(traces: &[Value]) -> (Vec<Span>, Vec<usize>) {
    let mut flat = Vec::new();
    let mut owner = Vec::new();
    for (ti, t) in traces.iter().enumerate() {
        for s in extract::find_reasoning_spans(t) {
            flat.push(s);
            owner.push(ti);
        }
    }
    (flat, owner)
}

fn write_prompts(dir: &Path) -> Result<()> {
    fs::write(dir.join("prompts/compress_lite.md"), prompts::compress_prompt(Level::Lite))?;
    fs::write(dir.join("prompts/compress_full.md"), prompts::compress_prompt(Level::Full))?;
    fs::write(dir.join("prompts/invert.md"),        prompts::invert_prompt())?;
    fs::write(dir.join("prompts/validate.md"),      prompts::validate_prompt())?;
    Ok(())
}

fn trace_id(trace: &Value, idx: usize) -> String {
    let raw = serde_json::to_string(trace).unwrap_or_default();
    let mut hasher = sha1_smol::Sha1::new();
    hasher.update(raw.as_bytes());
    let hex = hasher.digest().to_string();
    format!("t{:06}_{}", idx, &hex[..10])
}
